util/tracing/tracing.go

```diff
@@ -6,110 +6,128 @@ import (
 	"io"
 	"net/http"
 
-	"github.com/opentracing-contrib/go-stdlib/nethttp"
-	opentracing "github.com/opentracing/opentracing-go"
-	"github.com/opentracing/opentracing-go/ext"
-	"github.com/opentracing/opentracing-go/log"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/semconv"
+	"go.opentelemetry.io/otel/trace"
 )
 
 // StartSpan starts a new span as a child of the span in context.
 // If there is no span in context then this is a no-op.
-// The difference from opentracing.StartSpanFromContext is that this method
-// does not depend on global tracer.
-func StartSpan(ctx context.Context, operationName string, opts ...opentracing.StartSpanOption) (opentracing.Span, context.Context) {
-	parent := opentracing.SpanFromContext(ctx)
-	tracer := opentracing.Tracer(&opentracing.NoopTracer{})
+func StartSpan(ctx context.Context, operationName string, opts ...trace.SpanOption) (trace.Span, context.Context) {
+	parent := trace.SpanFromContext(ctx)
+	tracer := trace.NewNoopTracerProvider().Tracer("")
 	if parent != nil {
 		tracer = parent.Tracer()
-		opts = append(opts, opentracing.ChildOf(parent.Context()))
-	}
-	span := tracer.StartSpan(operationName, opts...)
-	if parent != nil {
-		return span, opentracing.ContextWithSpan(ctx, span)
 	}
+	ctx, span := tracer.Start(ctx, operationName, opts...)
 	return span, ctx
 }
 
 // FinishWithError finalizes the span and sets the error if one is passed
-func FinishWithError(span opentracing.Span, err error) {
+func FinishWithError(span trace.Span, err error) {
 	if err != nil {
-		fields := []log.Field{
-			log.String("event", "error"),
-			log.String("message", err.Error()),
-		}
+		span.RecordError(err)
 		if _, ok := err.(interface {
 			Cause() error
 		}); ok {
-			fields = append(fields, log.String("stack", fmt.Sprintf("%+v", err)))
+			span.SetAttributes(attribute.String(string(semconv.ExceptionStacktraceKey), fmt.Sprintf("%+v", err)))
 		}
-		span.LogFields(fields...)
-		ext.Error.Set(span, true)
+		span.SetStatus(codes.Error, err.Error())
 	}
-	span.Finish()
+	span.End()
 }
 
 // ContextWithSpanFromContext sets the tracing span of a context from other
 // context if one is not already set. Alternative would be
 // context.WithoutCancel() that would copy the context but reset ctx.Done
 func ContextWithSpanFromContext(ctx, ctx2 context.Context) context.Context {
 	// if already is a span then noop
-	if span := opentracing.SpanFromContext(ctx); span != nil {
+	if span := trace.SpanFromContext(ctx); span != nil {
```

Comment:
This needs to check if the returned span is valid as `trace.SpanFromContext()` will always return a non-nil value.